### PR TITLE
replaced javascript library for java lib plus homespun code

### PR DESCRIPTION
**Pull request by RowanErasmus**
As per our discussion #https://github.com/OHDSI/circe-be/issues/188 I've swapped out the javascript code for a pure java implementation. The java libraries available for semantic versioning don't have a readymade intersect function so I've written one myself (using some of the stuff the library did have)

There are some contraints, the original js library supported the full node semver syntax with tildes ~, carets ^ etc. for practical reasons this implementation only supports the basics as described in the Javadoc

To test the implementation I wrote a bunch of tests and checked the output of the new implementation against the old one, it appears to be working fine for the tests I conjured up.

If you can think of some kind of extra edge cases I did not test please let me know.

### CODE REVIEW

File: src/test/java/org/ohdsi/analysis/versioning/SemverUtilsTest.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    public void testGetRangesIntersectionWithRange1() {
+        String intersection = getRangesIntersection(asList(">=3.0.0", "<=5.0.0"));
+        assertEquals("3.0.0 - 5.0.0", intersection);
+    }
+
+    @Test
+    public void testGetRangesIntersectionWithRange2() {
+        String intersection = getRangesIntersection(asList("4 - 5", "<=5.0.0"));
+        assertEquals("4.0.0 - 5.0.0", intersection);
+    }
```
Comment (chrisknoll) [inline src/test/java/org/ohdsi/analysis/versioning/SemverUtilsTest.java:73]: Can you add a test that gets the intersection of `("4 - 5", "<5")`.   I don't think it can resolve in a range of 4.0.0 - 5.0.0 because 5.0 is not included and I wanted to understand how that will resolve.

File: pom.xml [hunk 2]
```diff
@@ -83,33 +80,10 @@
...
-      <artifactId>js-scriptengine</artifactId>
-      <version>${graalvm.version}</version>
-    </dependency>
-    <dependency>
-      <groupId>org.skyscreamer</groupId>
-      <artifactId>jsonassert</artifactId>
-      <version>1.5.1</version>
-      <scope>test</scope>
+      <groupId>com.vdurmont</groupId>
+      <artifactId>semver4j</artifactId>
```
Comment (chrisknoll) [inline pom.xml]: I believe the repo maintained by vdurmont is not maintained, but the github org `semver4j` is carrying it on. According to their README.md, this is how we reference in maven:

```
<dependency>
    <groupId>org.semver4j</groupId>
    <artifactId>semver4j</artifactId>
    <version>4.3.0</version>
</dependency>
```
Looks like it is up to version 4.3.0 under the semver4j org.

I think this has greater support for some of the npm syntax (not that we need it, just calling it out)

File: src/test/java/org/ohdsi/analysis/versioning/SemverUtilsTest.java (hunk 1 above)
Comment (RowanErasmus) [reply src/test/java/org/ohdsi/analysis/versioning/SemverUtilsTest.java:73]: You are right... it doesn't work, let me think about it a minute

Comment (chrisknoll) [reply src/test/java/org/ohdsi/analysis/versioning/SemverUtilsTest.java:73]: I might suggest that if you have an existing range like 4-5 that you then see a <5 come up, you can treat <X as <= (X-1).999.999 so that your range will get shrunk to 4-4.999.999.   Sounds hacky but I think it's going to be a typical thing where something gets deprecated in a major version and so you 'cap' the range at everything up to the current major.

Comment (RowanErasmus) [reply src/test/java/org/ohdsi/analysis/versioning/SemverUtilsTest.java:73]: I think the hacky workaround is probably better than a 'correct' but complex (buggy) implementation, I've made a change and added a test

File: pom.xml (hunk 2 above)
Comment (RowanErasmus) [reply pom.xml]: I've switched them round, this has its own perculiarites, but a bug in the other lib is fixed here so that is good :-)

**Comment by chrisknoll** [comment]
Ok, this is looking good to me, is there anything else you were trying to target or should I move forward with the PR?

File: src/test/java/org/ohdsi/analysis/versioning/SemverUtilsTest.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    @Test
+    public void testGetRangesIntersectionWithAstrix() {
+        String intersection = getRangesIntersection(asList("*", "<=5.0.0"));
+        assertEquals("<=5.0.0", intersection);
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void testGetRangesIntersectionWithEmptyList() {
+        getRangesIntersection(emptyList());
+    }
```
Comment (chrisknoll) [inline src/test/java/org/ohdsi/analysis/versioning/SemverUtilsTest.java:96]: There is another test that we should add that should result in an exception, which is the case where the range intersection results in nothing (i'm not sure if this should be excpetion or if it shoudl return null and if we need a cdm range somewhere we check for null):

```
getRangesIntersection(asList("<5",">6"));  /// should throw exception because this is a contradition or return NULL because there is no intersection.
```

Comment (RowanErasmus) [reply src/test/java/org/ohdsi/analysis/versioning/SemverUtilsTest.java:96]: This kind of result returned null, and the `CdmVersionUtils.java` then threw an exception. With a null as result the contract stays in tact. I have a test `testGetRangesIntersectionNoIntersection` for this already, I've added another one that is slightly different but also has no intersection.

**Comment by RowanErasmus** [comment]
For me this is fine, with the reduced package size CirceR should be able to get into CRAN and it removes a 'critical' vulnerability in Graal. Now there is probably a bug somewhere that will break something downstream, there always is when I write this kind of code, so I guess we should keep an eye out for that. Otherwise I'm all in favour of wrapping it up, also with Edwins updates it would be nice to get this further downstream :-)

**Comment by chrisknoll** [comment]
Ok, thanks, I'm going to pull this locally and run some of my Circe tests using this Util class (just to verify serialization) and I think we can be good to go.

File: src/test/java/org/ohdsi/analysis/versioning/SemverUtilsTest.java (hunk 3 above)
Comment (chrisknoll) [reply src/test/java/org/ohdsi/analysis/versioning/SemverUtilsTest.java:96]: Ok, side note, I think we will want CdmVersionUtils to just assign the null value instead of throwing an exception, because  it means that we can't even serialize a object with a cdm conflict in it (which I feel is too extreme) and instead we can have a validation for somethign like a cohort expresiso that if we have a null cdmVersion then that means there's a conflict in the expression.

not sure if we should do that as part of this PR or make a different issue out of it, but I can make a different issue out of it after this PR is done so we can just work off of this version of code for the fix.

Comment (RowanErasmus) [reply src/test/java/org/ohdsi/analysis/versioning/SemverUtilsTest.java:96]: That does make sense, I would however make it a separate issue, my intention with this change is to keep the functionality as much as it was and not change the behaviour in unexpected ways. Somebody downstream my be depending on that current hard failure to make sure their work is correct...

Comment (chrisknoll) [reply src/test/java/org/ohdsi/analysis/versioning/SemverUtilsTest.java:96]: Sounds good.

**Comment by chrisknoll** [comment]
I've updated the upstream master branch to make the next version 1.4.0-SNAPSHOT. can you sync your upstream fork and merge this PR with your forked master so you get the updated version?

**Comment by RowanErasmus** [comment]
@chrisknoll I've pulled in the changes and fixed the conflict

**Comment by chrisknoll** [comment]
There was 1 issue that i found that broke 2 tests: 

```
expected >=5.3.0 but was >=5.3
```

This is specifically related to [PayperPlanPeriod](https://github.com/OHDSI/circe-be/blob/1bec254c8606a40c4a60ccacc325124c5006aa84/src/main/java/org/ohdsi/circe/cohortdefinition/PayerPlanPeriod.java#L26) where we define this CDM version range:

```
@CdmVersion(range = ">=5.3")
```

So, I can think of 2 ways to fix this, maybe we have to use fully qualified semver major.minor.hotfix version syntax where we use this, OR we make the semver utlity return the version range in major.minor.hotfix form which means to replace 5.3 with 5.3.0 (or basically fill in zeroes for missing parts of the major.minor.hotfix). 

I feel like the latter is better option because it maintains backwards compatability about what the expected output is.   But if we don't want to change it in this repo, i'll have to update all the cdmVersion attributes in WebAPI to ensure they work off of a major.minor.hotfix format.

**Comment by chrisknoll** [comment]
Or, Option 3: change the test, if we want to say that the test was wrong to reject the >=5.3 and make the test consider >=5.3 the correct response.

**Comment by RowanErasmus** [comment]
Principally I would say if you provide >=5.3 and it returns >=5.3 then that is completely accurate, can't really argue with that :-)

But if a three part version is always expected downstream I guess that is what should be provided...

At the same time it seems that OHDSI currently prefers two part version numbers (3.0, 5.3, 5.4, 6.0) but in the past they also had a 5.3.1 version... As this is an OHDSI specific library it should probably conform to OHDSI preferred standards, but to be honest I'm a bit confused what those are :-)

All in all I don't really know what is the best way forward, maybe Anthony has some words of wisdom in todays meeting.

**Review by chrisknoll — APPROVED** [review]
